mleap-spark/src/main/scala/org/apache/spark/ml/bundle/ops/feature/StopWordsRemoverOp.scala

```diff
@@ -2,42 +2,35 @@ package org.apache.spark.ml.bundle.ops.feature
 
 import ml.combust.bundle.BundleContext
 import ml.combust.bundle.dsl._
-import ml.combust.bundle.op.{OpModel, OpNode}
-import org.apache.spark.ml.bundle.{ParamSpec, SimpleParamSpec, SimpleSparkOp, SparkBundleContext}
+import ml.combust.bundle.op.{OpModel}
+import org.apache.spark.ml.bundle.{MultiInOutFormatSparkOp, SimpleSparkOp, SparkBundleContext}
 import org.apache.spark.ml.feature.StopWordsRemover
 
 /**
   * Created by mikhail on 10/16/16.
   */
-class StopWordsRemoverOp extends SimpleSparkOp[StopWordsRemover] {
+class StopWordsRemoverOp extends SimpleSparkOp[StopWordsRemover] with MultiInOutFormatSparkOp[StopWordsRemover] {
   override val Model: OpModel[SparkBundleContext, StopWordsRemover] = new OpModel[SparkBundleContext, StopWordsRemover] {
     override val klazz: Class[StopWordsRemover] = classOf[StopWordsRemover]
 
     override def opName: String = Bundle.BuiltinOps.feature.stopwords_remover
 
     override def store(model: Model, obj: StopWordsRemover)
                       (implicit context: BundleContext[SparkBundleContext]): Model = {
-      model.withValue("stop_words", Value.stringList(obj.getStopWords)).
+      val result = model.withValue("stop_words", Value.stringList(obj.getStopWords)).
         withValue("case_sensitive", Value.boolean(obj.getCaseSensitive))
+      saveMultiInOutFormat(result, obj)
     }
 
     override def load(model: Model)
                      (implicit context: BundleContext[SparkBundleContext]): StopWordsRemover = {
-      new StopWordsRemover(uid = "").setStopWords(model.value("stop_words").getStringList.toArray).
+      val obj = new StopWordsRemover(uid = "").setStopWords(model.value("stop_words").getStringList.toArray).
         setCaseSensitive(model.value("case_sensitive").getBoolean)
+      loadMultiInOutFormat(model, obj)
     }
-
   }
 
   override def sparkLoad(uid: String, shape: NodeShape, model: StopWordsRemover): StopWordsRemover = {
-    new StopWordsRemover(uid = uid).setStopWords(model.getStopWords).setCaseSensitive(model.getCaseSensitive)
-  }
-
-  override def sparkInputs(obj: StopWordsRemover): Seq[ParamSpec] = {
-    Seq("input" -> obj.inputCol)
-  }
-
-  override def sparkOutputs(obj: StopWordsRemover): Seq[SimpleParamSpec] = {
-    Seq("output" -> obj.outputCol)
+    model
```

Comment:
should we still be doing ` new StopWordsRemover(uid = uid).setStopWords(model.getStopWords).setCaseSensitive(model.getCaseSensitive)`?

Reply:
(same idea for all 3 of these)

I had something like that at first, but this is kind of tricky and is incorrect/unnecessary.  How the code flow works.
1. We create the `StopWordsRemover` object in the `OpModel.load` method above this.  As part of this, the new `loadMultiInOutFormat` method sets the input/output parameters on that transformer object.
2. The `SimpleSparkOp.load` method calls `sparkLoad` with model equal to the output transformer object we created in step 1.  We do have the option of creating a new transformer object here, as long as we also would set the input/output parameters.  But this seems unnecessary and would result in having to duplicate all the multi in/out format multiplexing again?  So instead we just pass the model object though without changing it.
3.  The `SimpleSparkOp.load` method then calls `sparkInputs` and `sparkOutputs`.  For most transformers, `sparkInputs` and `sparkOutputs` are hardcoded since it just has to say what the transformers input/output parameters are called -- we can't hardcode because we need to support multi in/out formats.  The easiest way to handle this was to have the transformer object already set up with it's input/output parameters set up, which we did in step 1.
4. The `SimpleSparkOp.load` method then calls `SparkShapeLoader.loadShape` which has kind of a bad name and actually is responsible for setting the input/output parameters on the transformer.  For most transformers it would use the hardcoded parameter names provided by `sparkInputs` and `sparkOutputs`.  For the multi in/out format transformers it will just re-set the same parameter values we already set up.

So the key tricky thing is that for multi in/out formats we need to know the input/output parameters (step 3) before they are normally set (step 4).  Thus we need to set them up earlier outside of the usual code flow.  Step 1 was the logical choice since the serialization format can directly contain which format we're using.

I'm not really even sure why `sparkLoad` exists to be honest, but suspect there is some edge case transformer which needed to do something custom -- though am not sure why that custom logic couldn't be in step 1.  Or perhaps there is some historical reason that doesn't apply anymore?  In any case, most spark ops seemingly don't need the `sparkLoad` method at all and are creating a new transformer unnecessarily I think?